### ISSUE TITLE
Database hangs during shutdown.

### ISSUE THREAD

**Issue by tkroman** [neo4j/neo4j#4609]
I'm using neo4j 2.2.1.

99% of time the instance shuts down normally, but sometimes an application hangs in `db.shutdown()` and thread dump shows this:

```
"pool-8-thread-1" #84 prio=5 os_prio=0 tid=0x00007fe4dc00e800 nid=0x2175 waiting on condition [0x00007fe4b79f7000]
java.lang.Thread.State: TIMED_WAITING (sleeping)
at java.lang.Thread.sleep(Native Method)
at org.neo4j.kernel.impl.store.kvstore.RotationState$Rotation.rotate(RotationState.java:74)
at org.neo4j.kernel.impl.store.kvstore.RotationState$Rotation.rotate(RotationState.java:51)
at org.neo4j.kernel.impl.store.kvstore.AbstractKeyValueStore$RotationTask.rotate(AbstractKeyValueStore.java:278)
at org.neo4j.kernel.impl.store.kvstore.AbstractKeyValueStore$RotationTask.rotate(AbstractKeyValueStore.java:259)
at org.neo4j.kernel.impl.store.counts.CountsTracker.rotate(CountsTracker.java:141)
at org.neo4j.kernel.impl.store.NeoStore.flush(NeoStore.java:316)
at org.neo4j.kernel.impl.transaction.log.LogRotationControl.forceEverything(LogRotationControl.java:61)
at org.neo4j.kernel.NeoStoreDataSource.stop(NeoStoreDataSource.java:1116)
- locked <0x00000006d2eebd28> (a org.neo4j.kernel.impl.transaction.log.PhysicalWritableLogChannel)
- locked <0x00000006d2eeb258> (a org.neo4j.kernel.impl.transaction.log.PhysicalLogFile)
- locked <0x00000006d2844ce8> (a org.neo4j.kernel.NeoStoreDataSource)
at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.stop(LifeSupport.java:527)
at org.neo4j.kernel.lifecycle.LifeSupport.stop(LifeSupport.java:155)
- locked <0x00000006d3098280> (a org.neo4j.kernel.lifecycle.LifeSupport)
at org.neo4j.kernel.impl.transaction.state.DataSourceManager.stop(DataSourceManager.java:137)
at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.stop(LifeSupport.java:527)
at org.neo4j.kernel.lifecycle.LifeSupport.stop(LifeSupport.java:155)
- locked <0x00000006d272f818> (a org.neo4j.kernel.lifecycle.LifeSupport)
at org.neo4j.kernel.lifecycle.LifeSupport.shutdown(LifeSupport.java:185)
- locked <0x00000006d272f818> (a org.neo4j.kernel.lifecycle.LifeSupport)
at org.neo4j.kernel.InternalAbstractGraphDatabase.shutdown(InternalAbstractGraphDatabase.java:685)
```

**Comment by jexp** [comment]
@tkroman does it finish after that or just hang forever?

**Comment by tkroman** [comment]
No, it hangs forever. I was just repeating it to check out and found out these entries in `messages.log` (each time the database hangs, I get these):

```
2015-05-13 11:07:21.144+0000 INFO  [o.n.k.h.HighlyAvailableGraphDatabase]: Shutdown started
2015-05-13 11:07:21.145+0000 INFO  [o.n.k.h.HighAvailabilityConsoleLogger]: Write transactions to database disabled
2015-05-13 11:07:21.145+0000 INFO  [o.n.k.h.HighlyAvailableGraphDatabase]: Database is now unavailable
2015-05-13 11:07:21.146+0000 INFO  [o.n.k.h.c.m.MasterServer]: MasterServer communication server shutting down and unbinding from  ***
2015-05-13 11:07:21.194+0000 INFO  [o.n.k.h.c.HighAvailabilityModeSwitcher]: I am 1, moving to pending
2015-05-13 11:07:21.195+0000 INFO  [o.n.c.p.c.ClusterState]: Shutting down cluster: neo4j.ha
2015-05-13 11:07:21.196+0000 INFO  [o.n.k.h.HighAvailabilityConsoleLogger]: Instance 1 (this server)  left the cluster
2015-05-13 11:07:21.198+0000 INFO  [o.n.c.c.NetworkReceiver]: cluster://*** disconnected from me at cluster://***
2015-05-13 11:07:21.198+0000 INFO  [o.n.c.c.NetworkReceiver]: cluster://*** disconnected from me at cluster://***
2015-05-13 11:07:21.200+0000 INFO  [o.n.b.BackupServer]: BackupServer communication server shutting down and unbinding from  ***
2015-05-13 11:07:28.105+0000 INFO  [o.n.k.i.a.i.s.OnlineIndexSamplingJob]: Sampled index :IdProv(idProvOffset) with 1 unique values in sample of avg size 1 taken from index containing 1 entries
```

I'm talking about the last line (index sampling after database shutdown).

If I add `sleep(5000)` before the call to shutdown, it stops correctly. Looks like some other thread still tries to do something to the database, which sends it to the state it can't recover from.

**Comment by tkroman** [comment]
Yep, there it is: 

Sampler waits for `0x00000006d2c43c50`

```
"Neo4j HighlyAvailableGraphDatabase[/mp/eo4jFeaturesTest5878781630618007517]-1" #61 daemon prio=5 os_prio=0 tid=0x00007fc8edb56800 nid=0x4f0b waiting on condition [0x00007fc87d173000]
java.lang.Thread.State: WAITING (parking)
at sun.misc.Unsafe.park(Native Method)
- parking to wait for  <0x00000006d2c43c50> (a java.util.concurrent.locks.ReentrantReadWriteLock$FairSync)
at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireShared(AbstractQueuedSynchronizer.java:967)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireShared(AbstractQueuedSynchronizer.java:1283)
at java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.lock(ReentrantReadWriteLock.java:727)
at org.neo4j.kernel.impl.locking.LockWrapper.<init>(LockWrapper.java:40)
at org.neo4j.kernel.impl.locking.LockWrapper.readLock(LockWrapper.java:28)
at org.neo4j.kernel.impl.store.kvstore.AbstractKeyValueStore.updater(AbstractKeyValueStore.java:177)
at org.neo4j.kernel.impl.store.counts.CountsTracker.updateIndexCounts(CountsTracker.java:199)
at org.neo4j.kernel.impl.transaction.state.NeoStoreIndexStoreView.replaceIndexCounts(NeoStoreIndexStoreView.java:85)
at org.neo4j.kernel.impl.api.index.sampling.OnlineIndexSamplingJob.run(OnlineIndexSamplingJob.java:79)
at org.neo4j.kernel.impl.api.index.sampling.IndexSamplingJobTracker$1.run(IndexSamplingJobTracker.java:79)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
at org.neo4j.helpers.NamedThreadFactory$2.run(NamedThreadFactory.java:99)
```

And the main thread is waiting for `0x00000006d2c43c50` too:

```
"pool-8-thread-1" #82 prio=5 os_prio=0 tid=0x00007fc854013000 nid=0x4f20 waiting on condition [0x00007fc8539f7000]
java.lang.Thread.State: WAITING (parking)
at sun.misc.Unsafe.park(Native Method)
- parking to wait for  <0x00000006d2c43c50> (a java.util.concurrent.locks.ReentrantReadWriteLock$FairSync)
at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:870)
at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1199)
at java.util.concurrent.locks.ReentrantReadWriteLock$WriteLock.lock(ReentrantReadWriteLock.java:943)
at org.neo4j.kernel.impl.locking.LockWrapper.<init>(LockWrapper.java:40)
at org.neo4j.kernel.impl.locking.LockWrapper.writeLock(LockWrapper.java:33)
at org.neo4j.kernel.impl.store.kvstore.AbstractKeyValueStore.prepareRotation(AbstractKeyValueStore.java:206)
at org.neo4j.kernel.impl.store.counts.CountsTracker.rotate(CountsTracker.java:141)
at org.neo4j.kernel.impl.store.NeoStore.flush(NeoStore.java:316)
at org.neo4j.kernel.impl.transaction.log.LogRotationControl.forceEverything(LogRotationControl.java:61)
at org.neo4j.kernel.NeoStoreDataSource.stop(NeoStoreDataSource.java:1116)
- locked <0x00000006d2cbd4c8> (a org.neo4j.kernel.impl.transaction.log.PhysicalWritableLogChannel)
- locked <0x00000006d28ee4e0> (a org.neo4j.kernel.impl.transaction.log.PhysicalLogFile)
- locked <0x00000006d28ee618> (a org.neo4j.kernel.NeoStoreDataSource)
at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.stop(LifeSupport.java:527)
at org.neo4j.kernel.lifecycle.LifeSupport.stop(LifeSupport.java:155)
- locked <0x00000006d28ee7e0> (a org.neo4j.kernel.lifecycle.LifeSupport)
at org.neo4j.kernel.impl.transaction.state.DataSourceManager.stop(DataSourceManager.java:137)
at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.stop(LifeSupport.java:527)
at org.neo4j.kernel.lifecycle.LifeSupport.stop(LifeSupport.java:155)
- locked <0x00000006d2821b68> (a org.neo4j.kernel.lifecycle.LifeSupport)
at org.neo4j.kernel.lifecycle.LifeSupport.shutdown(LifeSupport.java:185)
- locked <0x00000006d2821b68> (a org.neo4j.kernel.lifecycle.LifeSupport)
at org.neo4j.kernel.InternalAbstractGraphDatabase.shutdown(InternalAbstractGraphDatabase.java:685)
```

**Comment by jexp** [comment]
Hey @tkroman sorry for now following up, this fell through the cracks. 
Are you able to reproduce this, and do you have the broken store still around? My colleague @lutovich would be very interested in that data.

**Comment by tkroman** [comment]
Hi Michael. No, I don't have the data. The issue somehow no longer presents itself (or I managed a workaround), so I did't return to it either.

**Comment by lutovich** [comment]
Hi @tkroman,
Thanks for the provided info, all this indeed looks like an issue that I'm chasing now.
First provided stacktrace was the only thing that we had until now.
I'll report progress in this thread.

**Comment by lutovich** [comment]
@tkroman is it possible that first shutdown issue occurred after a non-clean shutdown of the database?

It feels like first stacktrace and second pair of stacktraces represent different problems.
One represents store waiting infinitely for some transactions to complete and other probably represents deadlock when something got a write lock `0x00000006d2c43c50` and blocked sampler and main thread.

What kind of load do you put on the db before shutting it down? Maybe by chance you still have a full stacktrace for the second case?

Thanks!

**Comment by tkroman** [comment]
@lutovich I found an old piece of code that reproduces this issue for me. Here it is:

```
@Test
public void testInterruptBehaviour() throws Exception {
    Runnable r = () -> {
        CyclicBarrier barrier = new CyclicBarrier(2);
        AtomicReference<Thread> t = new AtomicReference<>();
        Future<?> f1 = Executors.newSingleThreadExecutor().submit(() -> {
            try {
                barrier.await();
                t.get().interrupt();
            } catch (Exception exn) {
                exn.printStackTrace();
            }
        });
        try {
            Executors.newSingleThreadExecutor().submit(() -> {
                sleep(100, TimeUnit.MILLISECONDS);
                Transaction tx = db().beginTx();
                db().createNode();
                tx.success();
                try {
                    t.set(Thread.currentThread());
                    barrier.await();
                    tx.close();
                } catch (Exception exn) {
                    exn.printStackTrace();
                }
            }).get();
        } catch (Exception exn) {
            exn.printStackTrace();
        }
        try {
            f1.get();
        } catch (Exception exn) {
            exn.printStackTrace();
        }
    };
    ExecutorService e = Executors.newFixedThreadPool(24);
    for (int i = 0; i < 24; i++) {
        e.execute(r);
    }
    e.shutdown();
    e.awaitTermination(1, TimeUnit.MINUTES);

    if (db().isAvailable(TimeUnit.MINUTES.toMillis(1))) {
        try (Transaction tx = db().beginTx()) {
            db().createNode(() -> "QWERTY").setProperty("sdf", 1241465423L);
            tx.success();
        }
        db().shutdown();
    }
}
```

There are few things worth mentioning:
- if I create a reasonably small (e.g. 4 instead of 24) amount of threads, the issue doesn't reproduce itself.
- if I rethrow exceptions, the issue doesn't reproduce itself
- I can't reproduce it in any other way than thrashing the database state

**Comment by lutovich** [comment]
Hi @tkroman,
I've tried to use provided test to reproduce the problem but it did not work for me. Executed on both JDK 7 & 8 on MacOS and Ubuntu (virtual box) in tight loop for hours but hanging did not show up.
At the same time your reference to index sampling was very helpful and one problem got the fix: https://github.com/neo4j/neo4j/pull/4818

Can you easily reproduce the problem with given test? Is it just my env that work differently?

**Comment by tkroman** [comment]
@lutovich Hi, yes, I can easily reproduce this. I assume you ran this test before you applied the fix from #4818? In that case, it's also environment-dependent. Is this the 100% same case as in the mentioned issue? Anyway, seeing as I'm always switching to the latest available neo4j version, I'll let you know if that didn't work for me, though I think it will.

FWIW, here's my Java version:

```
java version "1.8.0_45"
Java(TM) SE Runtime Environment (build 1.8.0_45-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.45-b02, mixed mode)
```

I have a 2-core (4 HT-threads) CPU / 16GB RAM / some SSD, in case that matters.

Did you try increasing the thread number (perhaps to some even more unreasonable values?)

**Comment by lutovich** [comment]
@tkroman I've executed the same test without fix on a virtual box with 4 cores. Will try to play with thread count. Thanks!

**Comment by lutovich** [comment]
Hi @tkroman,
Neo4j 2.2.4 was recently released. It should contain the mentioned fix. Please feel free to reopen this issue if the problem persists.
Thanks for help!

**Comment by tkroman** [comment]
Hi @lutovich,
Thanks for keeping me posted!

Will the fix be available in the next release of 2.3.x branch? We are still on M01 since M02 had some NDP dependencies issues, so it would be nice to have a fresh 2.3 release containing important fixes made lately (I'm talking about this one, deadlock detection-related one and some other major improvements we'd like to have). Is there any estimate for the next milestone?

**Comment by lutovich** [comment]
@tkroman yes, PR with fix was forward merged and so fix will be available in next 2.3 release. Unfortunately, I do not know when it'll happen but hope sooner rather than later :)

**Comment by amorgner** [comment]
I've got a 2.2.5 store which hangs on shutdown, jstack output is similar:

```
"Thread-13" #43 prio=5 os_prio=0 tid=0x00007fefb4001800 nid=0x369e waiting on condition [0x00007fef82f9a000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
    at java.lang.Thread.sleep(Native Method)
    at org.neo4j.kernel.impl.store.kvstore.RotationState$Rotation.rotate(RotationState.java:79)
    at org.neo4j.kernel.impl.store.kvstore.RotationState$Rotation.rotate(RotationState.java:50)
    at org.neo4j.kernel.impl.store.kvstore.AbstractKeyValueStore$RotationTask.rotate(AbstractKeyValueStore.java:281)
    at org.neo4j.kernel.impl.store.kvstore.AbstractKeyValueStore$RotationTask.rotate(AbstractKeyValueStore.java:262)
    at org.neo4j.kernel.impl.store.counts.CountsTracker.rotate(CountsTracker.java:146)
    at org.neo4j.kernel.impl.store.NeoStore.flush(NeoStore.java:317)
    at org.neo4j.kernel.impl.transaction.log.LogRotationControl.forceEverything(LogRotationControl.java:61)
    at org.neo4j.kernel.NeoStoreDataSource.stop(NeoStoreDataSource.java:1146)
    - locked <0x0000000726f00900> (a org.neo4j.kernel.impl.transaction.log.PhysicalWritableLogChannel)
    - locked <0x0000000724531760> (a org.neo4j.kernel.impl.transaction.log.PhysicalLogFile)
    - locked <0x0000000726faacb8> (a org.neo4j.kernel.NeoStoreDataSource)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.stop(LifeSupport.java:527)
    at org.neo4j.kernel.lifecycle.LifeSupport.stop(LifeSupport.java:155)
    - locked <0x00000007271b40b0> (a org.neo4j.kernel.lifecycle.LifeSupport)
    at org.neo4j.kernel.impl.transaction.state.DataSourceManager.stop(DataSourceManager.java:137)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.stop(LifeSupport.java:527)
    at org.neo4j.kernel.lifecycle.LifeSupport.stop(LifeSupport.java:155)
    - locked <0x000000071df953a0> (a org.neo4j.kernel.lifecycle.LifeSupport)
    at org.neo4j.kernel.lifecycle.LifeSupport.shutdown(LifeSupport.java:185)
    - locked <0x000000071df953a0> (a org.neo4j.kernel.lifecycle.LifeSupport)
    at org.neo4j.kernel.InternalAbstractGraphDatabase.shutdown(InternalAbstractGraphDatabase.java:698)
    at org.neo4j.server.database.LifecycleManagingDatabase.stop(LifecycleManagingDatabase.java:128)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.stop(LifeSupport.java:527)
    at org.neo4j.kernel.lifecycle.LifeSupport.stop(LifeSupport.java:155)
    - locked <0x000000071dd6abd8> (a org.neo4j.kernel.lifecycle.LifeSupport)
    at org.neo4j.server.AbstractNeoServer$6.run(AbstractNeoServer.java:616)
    at org.neo4j.helpers.RunCarefully.run(RunCarefully.java:47)
    at org.neo4j.server.AbstractNeoServer.stop(AbstractNeoServer.java:590)
    at org.neo4j.server.Bootstrapper$1.run(Bootstrapper.java:221)

   Locked ownable synchronizers:
    - None
```

@lutovich @jexp Does it make sense to reopen this issue?

**Comment by lutovich** [comment]
@amorgner It might be, yeah. Do you have any more information about this issue? Full stacktrace / heap dump / test case would be very appreciated.
